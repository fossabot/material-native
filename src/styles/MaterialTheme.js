'use strict';
import * as colors from './colors';
import shades from './shades';
const F = Object.freeze;

function normalizeTheme(theme) {
	return theme === 'dark' ? 'dark' : 'light';
}

function inverse(theme) {
	return theme === 'dark' ? 'light' : 'dark';
}

const themeTones = F({
	light: F({
		statusBar: colors.grey300,
		barStyle: 'dark-content',
		appBar: colors.grey100,
		background: colors.grey50,
		container: colors.white,
	}),
	dark: F({
		statusBar: colors.black,
		barStyle: 'light-content',
		appBar: colors.grey900,
		background: '#303030',
		container: colors.grey800,
	}),
});

export default class MaterialTheme {
	static defaultTheme = new MaterialTheme();

	constructor(settings) {
		this.settings = settings || {};
		// @todo recursively freeze settings
		this._applyDefaults();
		this._buildTheme();
		this._freeze();
	}

	_applyDefaults() {
		this.settings.theme = normalizeTheme(this.settings.theme);
	}

	_buildTheme() {
		const {theme} = this.settings;
		const inverseTheme = inverse(theme);

		this.themeTone = theme;
		this.inverseThemeTone = inverseTheme;

		this.palette = {
			primary: this.settings.primary || colors.indigo500,
			darkPrimary: this.settings.darkPrimary || colors.indigo700,
			accent: this.settings.accent || colors.pinkA200,
			background: themeTones[theme].background,
			container: themeTones[theme].container,
		};

		this.defaultToolbar = Object.assign({
			theme,
			background: themeTones[theme].appBar,
		}, this.settings.defaultToolbar || {});

		// @todo Automatically pick the right appBar theme based on the appBar background
		//       Many solid appBar colors should implicitly turn the theme within the appBar to dark
		const appBarTheme = normalizeTheme((this.settings.appBar && this.settings.appBar.theme) || this.defaultToolbar.theme);
		this.appBar = Object.assign({}, this.defaultToolbar, {
			theme: appBarTheme,
			background: this.palette.primary || this.defaultToolbar.background,
		}, this.settings.appBar || {});

		this.statusBar = Object.assign({
			background: this.palette.darkPrimary || themeTones[inverse(appBarTheme)].statusBar,
			translucentBackground: shades.dark.statusBar,
			barStyle: themeTones[inverse(appBarTheme)].barStyle,
		}, this.settings.statusBar || {});

		this.text = Object.assign({
			primaryColor: shades[inverseTheme].primaryText,
			secondaryColor: shades[inverseTheme].secondaryText,
			disabledColor: shades[inverseTheme].disabledText,
			hintColor: shades[inverseTheme].hintText,
			primaryOpacity: shades.opacity[inverseTheme].primaryText,
			secondaryOpacity: shades.opacity[inverseTheme].secondaryText,
			disabledOpacity: shades.opacity[inverseTheme].disabledText,
			hintOpacity: shades.opacity[inverseTheme].hintText,
		}, this.settings.text || {});

		this.icon = Object.assign({
			activeColor: shades[inverseTheme].activeIcon,
			inactiveColor: shades[inverseTheme].inactiveIcon,
			activeOpacity: shades.opacity[inverseTheme].activeIcon,
			inactiveOpacity: shades.opacity[inverseTheme].inactiveIcon,
		}, this.settings.icon || {});
	}

	_freeze() {
		Object.freeze(this); // @todo recursive
	}

	/**
	 * Create a new theme object that inherits settings from another
	 */
	extend(settings) {
		// @todo immutably merge `this.settings` with `settings` and pass that to `new MaterialTheme()`
	}
}

export function getMaterialTheme(settings) {
	return new MaterialTheme(settings);
}
