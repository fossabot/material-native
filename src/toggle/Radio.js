'use strict';
import React, {PureComponent} from 'react';
import {StyleSheet} from 'react-native';
import withMaterialTheme from '../styles/withMaterialTheme';
import CircleHighlight from '../touchable/CircleHighlight';
import CoreRadio from './CoreRadio';

/**
 * Standalone radio button with no label intended to be used alone with not linked to a parent touchable
 */
class Radio extends PureComponent {
	static defaultProps = {
		checked: false,
	};

	onPress = (e) => {
		this.props.onPress && this.props.onPress(e);
		this.props.onChangeChecked && this.props.onChangeChecked(!this.props.checked);
	};

	render() {
		const {
			onPress, // eslint-disable-line no-unused-vars
			style,
			...checkboxProps
		} = this.props;
		const {
			materialTheme,
			disabled,
			checked,
			accent,
			colorized: colorizedOverride,
		} = checkboxProps;

		const colorized = colorizedOverride || materialTheme.checkbox.colorized;
		const normalProps = materialTheme.checkbox.theme === 'dark' ? {light: true} : {dark: true};
		const tintProps = accent ? {accent: true} : {primary: true};

		const rippleProps = (
			(disabled && normalProps) ||
			(checked && tintProps) ||
			(colorized ? tintProps : normalProps)
		);

		// @todo Use disabled to disable ripple/press
		return (
			<CircleHighlight
				{...rippleProps}
				pointerEvents='box-only'
				style={[
					style,
					styles.root
				]}
				onPress={this.onPress}>
				<CoreRadio
					{...checkboxProps} />
			</CircleHighlight>
		);
	}
}

export default withMaterialTheme(Radio);

const styles = StyleSheet.create({
	root: {
		width: 48,
		height: 48,
		alignItems: 'center',
		justifyContent: 'center',
	},
});
