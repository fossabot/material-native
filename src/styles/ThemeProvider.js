'use strict';
import Observable from 'zen-observable';
import PropTypes from '../PropTypes';
import React, {PureComponent} from 'react';
import MaterialTheme from './MaterialTheme';

const defaultTheme = new MaterialTheme();

export default class ThemeProvider extends PureComponent {
	static contextTypes = {
		materialThemeObservable: PropTypes.Observable,
	};

	static childContextTypes = {
		materialThemeObservable: PropTypes.Observable.isRequired,
		memoizedMaterialStyles: PropTypes.func,
	};

	getChildContext() {
		return {
			materialThemeObservable: this.observable,
			memoizedMaterialStyles: this.memoizedMaterialStyles,
		};
	}

	componentWillMount() {
		// materialTheme
		if ( this.context.materialThemeObservable ) {
			this._parentSubscription = this.context.materialThemeObservable.subscribe({
				next: (theme) => {
					this.parentMaterialTheme = theme;
					this.handleThemeChange(this.props);
				},
				complete: () => {
					delete this._parentSubscription;
				},
			});
		} else {
			this.handleThemeChange(this.props);
		}
	}

	componentWillReceiveProps(nextProps) {
		// materialTheme
		this.handleThemeChange(nextProps);
	}

	componentWillUnmount() {
		if ( this._parentSubscription ) {
			this._parentSubscription.unsubscribe();
		}
	}

	// materialTheme
	observable = new Observable((observer) => {
		if ( this.materialTheme ) {
			observer.next(this.materialTheme);
		}

		this.observers.add(observer);

		return () => {
			this.observers.delete(observer);
		};
	});

	observers = new Set();

	materialTheme = undefined;
	parentMaterialTheme = undefined;

	// materialTheme
	handleThemeChange(props) {
		const baseTheme = props.theme || this.parentMaterialTheme || defaultTheme;

		// @todo Handle extend prop here

		if ( baseTheme !== this.materialTheme ) {
			this.materialTheme = baseTheme;
			this.observers.forEach((observer) => observer.next(this.materialTheme));
		}
	}

	// styleCache
	materialStyleCache = new Map();

	memoizedMaterialStyles = (materialTheme, mapThemeToStyles) => {
		let cache = this.materialStyleCache.get(mapThemeToStyles);
		if ( cache && cache.materialTheme === materialTheme ) {
			return cache.styles;
		}

		// @note If possible in the future, styles in `cache` can be unloaded at this point

		const cacheStyles = mapThemeToStyles(materialTheme);

		for ( const key in cacheStyles ) {
			Object.freeze(cacheStyles[key]);
		}
		Object.freeze(cacheStyles);

		cache = {
			materialTheme,
			styles: cacheStyles,
		};

		this.materialStyleCache.set(mapThemeToStyles, cache);

		return cache.styles;
	};

	// boilerplate
	render() {
		return React.Children.only(this.props.children);
	}
}
