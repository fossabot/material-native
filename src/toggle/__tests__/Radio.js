'use strict';
import React from 'react';
import MaterialTheme from '../../styles/MaterialTheme';
import ThemeProvider from '../../styles/ThemeProvider';
import Radio from '../Radio';

import ReactTestRenderer from 'react-test-renderer';
import ToJSX from 'react-element-to-jsx-string';

const theme = new MaterialTheme({});

/* eslint-disable react/jsx-key */
const snapshotTests = [
	<Radio />,
	<Radio onLayout={() => {}} />,
	<Radio checked />,
	<Radio disabled />,
	<Radio accent checked />
];
/* eslint-enable react/jsx-key */

for ( const root of snapshotTests ) {
	test(ToJSX(root, {maxInlineAttributesLineLength: Infinity}), () => {
		const host = ReactTestRenderer.create(
			<ThemeProvider theme={theme}>
				{root}
			</ThemeProvider>
		);
		expect(host.toJSON()).toMatchSnapshot();
	});
}

test('<Radio /> to have measure methods', () => {
	let instance;
	ReactTestRenderer.create(
		<ThemeProvider theme={theme}>
			<Radio ref={(ref) => instance = ref} />
		</ThemeProvider>
	);

	expect(instance).toBeDefined();
	expect(instance.measure).toBeInstanceOf(Function);
	expect(instance.measureInWindow).toBeInstanceOf(Function);
	expect(instance.measureLayout).toBeInstanceOf(Function);
});
