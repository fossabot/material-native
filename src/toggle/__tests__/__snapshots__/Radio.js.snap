// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<withMaterialTheme(Radio) /> 1`] = `
<View
  accessible={true}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  pointerEvents="box-only"
  rippleLocation="center"
  style={
    Array [
      Array [
        undefined,
        Object {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "width": 48,
        },
      ],
      Object {
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
      },
    ]
  }
>
  <View
    style={
      Array [
        undefined,
        Object {
          "alignItems": "center",
          "height": 24,
          "justifyContent": "center",
          "width": 24,
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.54)",
            "fontSize": 24,
          },
          Array [
            undefined,
            undefined,
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
  </View>
  <View
    pointerEvents="none"
    style={
      Array [
        Object {
          "overflow": "hidden",
          "position": "absolute",
        },
        Object {
          "borderRadius": 0,
          "height": 1,
          "width": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 0,
          "height": 1,
          "left": -0,
          "opacity": 0,
          "position": "absolute",
          "top": -0,
          "width": 1,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 0,
          "height": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "transform": Array [
            Object {
              "scale": 0,
            },
          ],
          "width": 0,
        }
      }
    />
  </View>
</View>
`;

exports[`<withMaterialTheme(Radio) accent checked /> 1`] = `
<View
  accessible={true}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  pointerEvents="box-only"
  rippleLocation="center"
  style={
    Array [
      Array [
        undefined,
        Object {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "width": 48,
        },
      ],
      Object {
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
      },
    ]
  }
>
  <View
    style={
      Array [
        undefined,
        Object {
          "alignItems": "center",
          "height": 24,
          "justifyContent": "center",
          "width": 24,
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#FF4081",
            "fontSize": 24,
          },
          Array [
            undefined,
            undefined,
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
  </View>
  <View
    pointerEvents="none"
    style={
      Array [
        Object {
          "overflow": "hidden",
          "position": "absolute",
        },
        Object {
          "borderRadius": 0,
          "height": 1,
          "width": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#FF4081",
          "borderRadius": 0,
          "height": 1,
          "left": -0,
          "opacity": 0,
          "position": "absolute",
          "top": -0,
          "width": 1,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "#FF4081",
          "borderRadius": 0,
          "height": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "transform": Array [
            Object {
              "scale": 0,
            },
          ],
          "width": 0,
        }
      }
    />
  </View>
</View>
`;

exports[`<withMaterialTheme(Radio) checked /> 1`] = `
<View
  accessible={true}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  pointerEvents="box-only"
  rippleLocation="center"
  style={
    Array [
      Array [
        undefined,
        Object {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "width": 48,
        },
      ],
      Object {
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
      },
    ]
  }
>
  <View
    style={
      Array [
        undefined,
        Object {
          "alignItems": "center",
          "height": 24,
          "justifyContent": "center",
          "width": 24,
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#3F51B5",
            "fontSize": 24,
          },
          Array [
            undefined,
            undefined,
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
  </View>
  <View
    pointerEvents="none"
    style={
      Array [
        Object {
          "overflow": "hidden",
          "position": "absolute",
        },
        Object {
          "borderRadius": 0,
          "height": 1,
          "width": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#3F51B5",
          "borderRadius": 0,
          "height": 1,
          "left": -0,
          "opacity": 0,
          "position": "absolute",
          "top": -0,
          "width": 1,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "#3F51B5",
          "borderRadius": 0,
          "height": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "transform": Array [
            Object {
              "scale": 0,
            },
          ],
          "width": 0,
        }
      }
    />
  </View>
</View>
`;

exports[`<withMaterialTheme(Radio) disabled /> 1`] = `
<View
  accessible={true}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  pointerEvents="box-only"
  rippleLocation="center"
  style={
    Array [
      Array [
        undefined,
        Object {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "width": 48,
        },
      ],
      Object {
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
      },
    ]
  }
>
  <View
    style={
      Array [
        undefined,
        Object {
          "alignItems": "center",
          "height": 24,
          "justifyContent": "center",
          "width": 24,
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.38)",
            "fontSize": 24,
          },
          Array [
            undefined,
            undefined,
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
  </View>
  <View
    pointerEvents="none"
    style={
      Array [
        Object {
          "overflow": "hidden",
          "position": "absolute",
        },
        Object {
          "borderRadius": 0,
          "height": 1,
          "width": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 0,
          "height": 1,
          "left": -0,
          "opacity": 0,
          "position": "absolute",
          "top": -0,
          "width": 1,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 0,
          "height": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "transform": Array [
            Object {
              "scale": 0,
            },
          ],
          "width": 0,
        }
      }
    />
  </View>
</View>
`;
